src/main/scala/beam/utils/MathUtils.scala

```diff
@@ -162,6 +162,40 @@ object MathUtils {
   def nanToZero(x: Double) = if (x.isNaN) { 0.0 }
   else { x }
 
+  /**
+    * It selects random elements out of a collection.
+    * It is designed to be performant on not indexed collections (like Sets).
+    * @param xs the collection
+    * @param n number of elements to select
+    * @param random a Random
+    * @tparam T type of elements
+    * @return an array of selected elements
+    */
+  def selectRandomElements[T: ClassTag](xs: Iterable[T], n: Int, random: Random): Array[T] = {
```

Comment:
I made comments on areas of this particular implementation, however I would be interested in why not something akin to:

```
val indexedIterable = xs.toVector
val toSelect = scala.collection.mutable.Set[Int]() 
val maxSize = xs.size - 1
while(toSelect.size < n){
  toSelect += random.nextInt(maxSize)
}
toSelect.map(indexedIterable)
```

Although I would leave it as `scala.util.Random` instead of taking in a `random`, as it is already fairly random that way.

Also, you can optimize the edge cases here, but I think it is more efficient in most cases - you pay a one time index and then move on.